rt/src/socket/socket_address.rs

```diff
@@ -2,40 +2,28 @@ use socket2::SockAddr;
 
 #[cfg(unix)]
 use {
-    libc::sockaddr_un,
-    std::ffi::OsStr,
-    std::mem::transmute,
-    std::os::{raw::c_char, unix::ffi::OsStrExt},
+    rustix::net::SocketAddrUnix, std::ffi::OsStr, std::os::unix::ffi::OsStrExt,
 };
 
-#[cfg(unix)]
-#[cfg_attr(feature = "cargo-clippy", allow(clippy::uninit_assumed_init))]
-fn sun_path_offset(addr: &sockaddr_un) -> usize {
-    let base = addr as *const sockaddr_un as usize;
-    let path = &addr.sun_path as *const c_char as usize;
-
-    path - base
-}
-
 #[cfg(unix)]
 fn unix_socket_path(sockaddr: &SockAddr) -> String {
-    let raw_addr = unsafe { &*(sockaddr.as_ptr() as *const sockaddr_un) };
-    let len = sockaddr.len() as usize - sun_path_offset(raw_addr);
-    let path = unsafe { transmute::<&[c_char], &[u8]>(&raw_addr.sun_path) };
-
-    if len == 0 || (cfg!(not(target_os = "linux")) && raw_addr.sun_path[0] == 0)
-    {
+    let sock_path = match sockaddr.as_pathname() {
+        Some(v) => v,
+        None => return String::new(),
+    };
+    let raw_addr = SocketAddrUnix::new(sock_path)
+        .map_err(|_| String::new())
+        .expect("SocketAddrUnix::new");
```

Comment:
The `expect` here makes the `map_err` redundant, because the error isn't ever used.

More importantly, the `expect` shouldn't be here in the first place, as we should instead just return the empty `String` in case of the failure.

The following should be correct (ignoring formatting for the sake of this example, but rustfmt should take care of that for you):

```rust
let path = if let Some(addr) = sockaddr.as_pathname().and_then(|v| SocketAddrUnix::new(v).ok()).and_then(|v| v.path()) {
  path
} else {
  return String::new();
}
```

The `cfg!` is no longer needed, as looking at the [Rustix code](https://docs.rs/rustix/latest/src/rustix/backend/linux_raw/net/addr.rs.html#70-81) of `SocketAddrUnix::path`, it already returns `None` if the first byte is a NULL byte. That means this snippet should be enough to replace everything leading up to the `OsStr::from_bytes` line below.

Reply:
I'm getting this issue if I use `and_then` syntax as your suggestion:
![image](https://github.com/inko-lang/inko/assets/5485589/7da1db89-a022-4896-adbd-fde6e7f80dd3)

Reply:
The error occurred because I call `a.path()`, it returns a reference to data inside `a`, but `a` is a temp value that will be dropped at the end of the `and_then` closure, it would lead to a dangling reference.

I'm changing this to another version, lemme push a fix and you can review this again.